edgraph/server.go

```diff
@@ -1883,6 +1884,356 @@ func (s *Server) CheckVersion(ctx context.Context, c *api.Check) (v *api.Version
+				},
+			},
+		}},
+		CommitNow: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if resp.Metrics.NumUids["q"] != 1 {
+		return nil, errors.Errorf("namespace [%v] not found", in.FromNs)
+	}
+
+	glog.Infof("Renamed namespace [%v] to [%v]", in.FromNs, in.ToNs)
+	return &apiv25.RenameNamespaceResponse{}, nil
+}
+
+func (s *Server) ListNamespaces(ctx context.Context, in *apiv25.ListNamespacesRequest) (
+	*apiv25.ListNamespacesResponse, error) {
+
+	if err := AuthGuardianOfTheGalaxy(ctx); err != nil {
+		s := status.Convert(err)
+		return nil, status.Error(s.Code(),
+			"Non guardian of galaxy user cannot list namespaces. "+s.Message())
+	}
+
+	const q = `{q(func: type(dgraph.namespace)) {
+			dgraph.namespace.name
+			dgraph.namespace.id }}`
+	resp, err := (&Server{}).doQuery(x.AttachJWTNamespace(ctx),
+		&Request{req: &api.Request{Query: q}, doAuth: NoAuthorize})
+	if err != nil {
+		return nil, err
+	}
+	var data struct {
+		NsList []struct {
+			ID   uint64 `json:"dgraph.namespace.id"`
+			Name string `json:"dgraph.namespace.name"`
+		} `json:"q"`
+	}
+	if err := json.Unmarshal(resp.GetJson(), &data); err != nil {
+		return nil, err
+	}
+	dataNsList := make(map[uint64]string)
+	for _, e := range data.NsList {
+		dataNsList[e.ID] = e.Name
+	}
+
+	schNsList := schema.State().Namespaces()
+	result := &apiv25.ListNamespacesResponse{NsList: make(map[string]*apiv25.Namespace)}
+	for id := range schNsList {
+		if name, ok := dataNsList[id]; !ok {
+			name = fmt.Sprintf("dgraph-%d", id)
+			result.NsList[name] = &apiv25.Namespace{Name: name, Id: id}
+		} else {
+			result.NsList[name] = &apiv25.Namespace{Name: name, Id: id}
+		}
+	}
+
+	return result, nil
+}
+
+// verifyNamespaceName ensures that name only has alpha numeric characters as
+// well as underscores and hyphens. It also ensures that the name is not empty.
+func verifyNamespaceName(name string) error {
+	if name == "" {
+		return errors.Errorf("namespace name cannot be empty")
+	}
+	hasInvalidChars := strings.ContainsFunc(name, func(r rune) bool {
+		return !unicode.IsLetter(r) && !unicode.IsNumber(r) && r != '_' && r != '-'
+	})
+	if hasInvalidChars {
+		return fmt.Errorf("namespace name [%v] has invalid characters", name)
+	}
+	if strings.HasPrefix(name, "_") || strings.HasPrefix(name, "-") {
+		return fmt.Errorf("namespace name [%v] cannot start with _ or -", name)
+	}
+	if strings.HasPrefix(name, "dgraph") {
+		return fmt.Errorf("namespace name [%v] cannot start with dgraph", name)
+	}
+	if name == "system" || name == "galaxy" {
+		return fmt.Errorf("namespace name [%v] is reserved", name)
+	}
+	if _, err := strconv.ParseInt(name, 10, 64); err == nil {
+		return fmt.Errorf("namespace name [%v] cannot be a number", name)
+	}
+	return nil
+}
+
+func verifyDropNamespaceName(name string) error {
+	if name == "" {
+		return errors.Errorf("namespace name cannot be empty")
+	}
+	if name == "system" || name == "galaxy" || name == "dgraph-0" {
```

Comment:
what is dgraph-0. Is it being used in hypermode? If so can we make some enum out of it so that it's easy to keep track

Reply:
That is the convention we are going to follow for pre v25 namespaces. Read here https://www.notion.so/hypermodeinc/Preview-Dgraph-v25-19de0dd4f736808bb1d7ffe98d38a3c8

Reply:
Lets create the enum?